### ISSUE TITLE
Revalidate value ignored on page.tsx until first revalidation

### ISSUE THREAD

**Issue by JonBernardHealia** [vercel/next.js#54460]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
npx --no-install next info

    Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.6.0: Wed Jul  5 22:22:05 PDT 2023; root:xnu-8796.141.3~6/RELEASE_ARM64_T6000
    Binaries:
      Node: 16.17.1
      npm: 8.15.0
      Yarn: 1.22.19
      pnpm: N/A
    Relevant Packages:
      next: 13.4.20-canary.3
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 5.1.3
    Next.js Config:
      output: N/A
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

App Router, Routing (next/router, next/navigation, next/link)

### Link to the code that reproduces this issue or a replay of the bug

https://github.com/JonBernardHealia/next-13-caching-issue

### To Reproduce

We've added Prisma to demonstrate interacting with a database.

### Setup/start app

1. Install `yarn` globally if you don't have it.
2. Run `yarn` in install packages
3. Start/Install Docker Desktop
4. Run `yarn db` to start postgres db in docker
5. In a new terminal, run `yarn migrate` to set up the db
6. Kill `yarn db` process
7. Run `yarn dev` to start app
8. Go to `localhost:3000` in your browser

### Steps to reproduce

1. Notice no todos show in list
2. Click `Add to-do` link at the top
3. On the add page, click the `add todo` button several times. Notice the client request counter goes up.
4. Click `Back to list` link at the top
5. Notice no todos show in list
6. Click the add and back links several times and no items in the list show up until the cache is revalidated once
7. When they do show up, add some more todos and click the back link. the new todos show up instantaneously

Note: refreshing the browser updates the list every time (but restarts that first revalidation clock), just not when you navigate with the router through the `Link` component or the `useRouter` hook from `next/navigation`.

### Describe the Bug

It appears that the `revalidate` value on pages in the new App router (`app/page.tsx`) is ignored until the first cache is invalidated. Once that happens, updates happen instantaneously.

### Expected Behavior

I would expect that if you use `export const revalidate = 0;` in a server component, it would never cache what is on this page, but it appears to be doing that on first render.

**Comment by JonBernardHealia** [comment]
Mentioning #42290 and #49417 because it looks like no one could create a reproduction repo for this issue.

**Comment by okanji** [comment]
I have the exact same issue and have not been able to find a solution. I have tried:

```js
export const dynamic = "force-dynamic";
export const revalidate = 0;
```

at the top of the server component but it does not work.

**Comment by okanji** [comment]
@JonBernardHealia , have you managed to find a solution ?

**Comment by JonBernardHealia** [comment]
> @JonBernardHealia , have you managed to find a solution ?

No, I haven't. The only workaround I've been able to find is to ditch the `next/nagigation` router entirely by using `window.local.href = '...'` or bypassing caching entirely by making API calls when you first load that page.

**Comment by JonBernardHealia** [comment]
I did find this workaround. 

https://www.reddit.com/r/nextjs/comments/14nttyp/how_to_handle_caching_with_user_specific_data_app/

Adding `router.refresh()` after specific changes will clear out all browser-side next js caching. Probably not the best solution, but it can help.

**Comment by okanji** [comment]
Great thanks
